File: DS _LAB_SEM_3/Stack_Program/Cir_Queue.cpp
#include<stdio.h>

void insert(void);
void display(void);
void del(void);

int a[5];
int front=-1;
int rear=-1;

main()
{
	int ch;
	
	do
	{
		printf("\n Enter 0 for Exist");
		printf("\n Enter 1 for insert");
		printf("\n Enter 2 for display");
		printf("\n Enter 3 for delete");
		
		printf("\n Enter choice: ");
		scanf("%d",&ch);
		
		switch(ch)
		{
			case 0:
					exit(1);
	
			case 1:
					insert();
					break;
			case 2:
					display();
					break;
			case 3:
					del();
					break;
			default:
				printf("Enter Valid Number");
		}
	}while(ch!=0);
	
	getch();
}

void insert(void)
{
	if(front==0 && rear==4)
	{
		printf("\n Circular Queue is Full ");
	}
	else if(rear==front-1)
	{
		printf("\n Circular Queue is Full ");
	}
	else
	{
		if(front==-1 && rear==-1)
		{
			front++;
			rear++;
		}
		else if(front>0 && rear==4)
		{
			rear=0;
		}
		else
		{
			rear++;
		}
		printf("\n Enter Number: ");
		scanf("%d",&a[rear]);
	}
	getch();
}

void display(void)
{
	int i;
	if(front==-1 && rear==-1)
	{
		printf("\n Queue is Empty");
	}
	else
	{
		if(rear<front)
		{
			for(i=front;i<=4;i++)
			{
				printf("%d",a[i]);
			}
			for(i=0;i<=rear;i++)
			{
				printf("%d",a[i]);
			}
		}
		else
		{
			for(i=front;i<=rear;i++)
			{
				printf("\n %d",a[i]);
			}
		}
	}
	getch();
}

void del(void)
{
	if(front==-1 && rear==-1)
	{
		printf("\n Circular Queue is Empty");
	}
	else
	{
		if(front==rear)
		{
			front=-1;
			rear=-1;
		}
		else if(front==0 && rear>=0)
		{
			front=0;
		}
		else
		{
			front++;
		}
	}
	getch();
}

File: DS _LAB_SEM_3/Exam Practice/Cir_Queue.java
import java.util.Scanner;

class CirQueueMain
{
    int size,front,rear;
    int [] queue;

    CirQueueMain(int size)
    {
        queue = new int[size];
        this.size = size;
        front = rear = -1;
    }

    public void insert(int n)
    {
        if(front ==0 && rear == size-1)
        {
            System.out.println("Queue Overflow");
        }
        else if(rear == front-1)
        {
            System.out.println("Queue Overflow");
        }
        else
        {
            if(front==-1 && rear==-1)
            {
                front = rear = 0;
            }
            else if(front>0 && rear == size-1)
            {
                rear = 0;
            }
            else
            {
                rear++;
            }
            queue[rear] = n;
        }
    }

    public int del()
    {
        if(front == -1 && rear == -1)
        {
            System.out.println("Circular Queue Is Empty");
            return 0;
        }
        else
        {
            int x = queue[front];
            if(front==rear)
            {
                front = rear = 0;
            }
            else if(front == size-1)
            {
                front = 0;
            }
            else
            {
                front++;
            }
            return x;
        }
    }

    public void display() 
    {
        if (front == -1 && rear == -1) 
        {
            System.out.println("Circular Queue Is Empty");
        }
        else 
        {
            if (rear < front) 
            {
                for (int i = front; i < size; i++) 
                {
                    System.out.print(queue[i] + " ");
                }
                for (int i = 0; i <= rear; i++) 
                {
                    System.out.print(queue[i] + " ");
                }
            } 
            else 
            {
                for (int i = front; i <= rear; i++) 
                {
                    System.out.print(queue[i] + " "); 
                }
            }
        System.out.println(); 
        }
    }

    public void count()
    {
        if(front==-1 && rear==-1)
        {
            System.out.println("Circular Queue Is Empty");
        }
        else
        {
            int count;

            if(front<=rear)
            {
                count = rear - front + 1;
            }
            else
            {
                count = size - front + rear + 1;
            }
            System.out.println("Total Nodes of Elemetn in Cirucular Queue:- "+count);
        }
    }
}
public class Cir_Queue 
{
   public static void main(String[] args) 
   {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter The size of Cirucular Queue:-");
        int size = sc.nextInt();

        CirQueueMain qi = new CirQueueMain(size);

        while(true)
        {
            System.out.println("Enter 1 for Insert Element in Queue:");
            System.out.println("Enter 2 for Delete Element in Queue:");
            System.out.println("Enter 3 for Count  Element in Queue:");
            System.out.println("Enter 4 for Display Element in Queue:");
            System.out.println("Enter 5 for exist.....");

            System.out.println("Enter Your Choice:-");
            int ch = sc.nextInt();

            switch(ch)
            {
                case 1:
                        System.out.println("Enter Element In Queue:- ");
                        int n = sc.nextInt();
                        qi.insert(n);
                        break;
                
                case 2:
                        qi.del();
                        break;
                
                case 3:
                        qi.count();
                        break;
                
                case 4:
                        qi.display();
                        break;

                case 5:
                        System.exit(0);
                        break;
                default:
                        System.out.println("Enter Valid Number in for Search....");
            }
            System.out.println("");
            System.out.println();
        }  
    }
} 